packages/subcommands/src/lib/utils/SubCommandsHandler.ts

```diff
@@ -0,0 +1,152 @@
+import { SubCommandNotLoadedError } from '../errors/SubCommandNotLoadedError';
+import {
+	CommandStore,
+	IPreconditionContainer,
+	PreCommandRunPayload,
+	PreconditionContainerArray,
+	PreconditionContainerSingle
+} from '@sapphire/framework';
+import { SubCommandMissingOptionsError } from '../errors/SubCommandMissingOptionsError';
+
+export type ResolveSubCommandPayload = Omit<PreCommandRunPayload, 'context'>;
+
+export interface SubCommandsPluginOptions {
+	/**
+	 * Preconditions that will be overlap in parents by their sub commands
+	 * @since 0.0.1
+	 * @default []
+	 */
+	overlappedPreconditions?: string[];
+}
+
+export const enum SubCommandsEvents {
+	PreSubCommandRun = 'preSubCommandRun',
+	SubCommandDenied = 'subCommandDenied',
+	SubCommandAccepted = 'subCommandAccepted',
+	SubCommandRun = 'subCommandRun',
+	SubCommandSuccess = 'subCommandSuccess',
+	SubCommandError = 'subCommandError',
+	SubCommandFinish = 'subCommandFinish'
+}
+
+export class SubCommandsHandler {
+	public readonly options: Required<SubCommandsPluginOptions>;
+
+	/**
+	 * @since 0.0.1
+	 * @param options Optional Subcommands plugin settings.
+	 */
+	public constructor(options: SubCommandsPluginOptions = {}) {
+		this.options = {
+			overlappedPreconditions: options.overlappedPreconditions ?? []
+		};
+	}
+
+	/**
+	 * Resolves sub command by passed payload.
+	 * @param payload The params that need for resolving sub command.
+	 */
+	public async resolveSubCommand({ command, message, parameters }: ResolveSubCommandPayload) {
+		const args = await command.preParse(message, parameters);
+		const store = command.store as CommandStore;
+
+		let nameFromArgs = await args.pickResult('string');
+		let iterableCommand = command;
+		let method = undefined;
+		let parentName = iterableCommand.name;
+		do {
+			if (!nameFromArgs.success || !iterableCommand.subCommands) break;
+
+			const soughtName = nameFromArgs.value.toLowerCase();
+			const subCommandInfo = iterableCommand.subCommands.find((s) => s.name === soughtName);
+			if (!subCommandInfo) break;
+
+			if (!subCommandInfo.command && !subCommandInfo.method) {
+				throw new SubCommandMissingOptionsError(subCommandInfo.name, parentName);
+			}
+
+			const soughtCommand = store.get(subCommandInfo.command ?? '') ?? (subCommandInfo.method ? command : undefined);
+			if (!soughtCommand) break;
+
+			parentName = soughtCommand.name;
+			iterableCommand = soughtCommand;
+			method = subCommandInfo.method;
+			args.save();
+
+			if (command === iterableCommand && subCommandInfo.method) break;
+		} while ((nameFromArgs = await args.pickResult('string')));
```

Comment:
Remind me, @kyranet, will this run again or will it first run the do block, then the while condition

Reply:
@vladfrangu Not actually realize what do u mean

Reply:
I was asking kyra if, in its current state, nameFromArgs would be re-fetched first OR it'll use the current nameFromArgs then re-fetch

Reply:
Please use `args.next()`.